libs/vkd3d/memory.c

```diff
@@ -659,9 +756,303 @@ void vkd3d_memory_allocator_cleanup(struct vkd3d_memory_allocator *allocator, st
 
     vkd3d_free(allocator->chunks);
     vkd3d_va_map_cleanup(&allocator->va_map);
+    vkd3d_memory_allocator_cleanup_clear_queue(allocator, device);
     pthread_mutex_destroy(&allocator->mutex);
 }
 
+static HRESULT vkd3d_memory_allocator_flush_clears_locked(struct vkd3d_memory_allocator *allocator,
+        struct d3d12_device *device)
+{
+    const VkPipelineStageFlags vk_stage_mask = VK_PIPELINE_STAGE_ALL_COMMANDS_BIT;
+    struct vkd3d_memory_clear_queue *clear_queue = &allocator->clear_queue;
+    const struct vkd3d_vk_device_procs *vk_procs = &device->vk_procs;
+    VkTimelineSemaphoreSubmitInfoKHR timeline_info;
+    struct vkd3d_queue *queue, *internal_queue;
+    VkCommandBufferBeginInfo begin_info;
+    VkMemoryBarrier memory_barrier;
+    VkCommandBuffer vk_cmd_buffer;
+    VkCommandPool vk_cmd_pool;
+    VkSubmitInfo submit_info;
+    uint32_t queue_mask;
+    VkQueue vk_queue;
+    VkResult vr;
+    size_t i;
+
+    if (!clear_queue->allocations_count)
+        return S_OK;
+
+    /* Record commands late so that we can simply remove allocations from
+     * the queue if they got freed before the clear commands got dispatched,
+     * rather than rewriting the command buffer or dispatching the clear */
+    internal_queue = device->queues[VKD3D_QUEUE_FAMILY_INTERNAL_COMPUTE];
+    vk_cmd_buffer = clear_queue->vk_command_buffers[clear_queue->command_buffer_index];
+    vk_cmd_pool = clear_queue->vk_command_pools[clear_queue->command_buffer_index];
+
+    if ((vr = VK_CALL(vkResetCommandPool(device->vk_device, vk_cmd_pool, 0))))
```

Comment:
Need to wait for timeline to complete before we can reset a command pool.

Reply:
Fixed.